prov/efa/src/efa_av.c

```diff
@@ -154,8 +145,198 @@ static inline int efa_av_is_valid_address(struct efa_ep_addr *addr)
+{
+	int err;
+#if ENABLE_DEBUG
+	struct efa_ah *tmp;
+
+	HASH_FIND(hh, av->ah_map, ah->gid, EFA_GID_LEN, tmp);
+	assert(tmp == ah);
+#endif
+	assert(ah->used > 0);
+	ah->used -= 1;
+	if (ah->used == 0) {
+		HASH_DEL(av->ah_map, ah);
+		err = ibv_destroy_ah(ah->ibv_ah);
+		if (err)
+			EFA_WARN(FI_LOG_AV, "ibv_destroy_ah failed! err=%d\n", err);
+		free(ah);
+	}
+}
+
+static
+void efa_conn_release(struct efa_av *av, struct efa_conn *conn);
+
+/**
+ * @brief initialize the rdm related resources of an efa_conn object
+ *
+ * This function setup rdm_peer and shm address for an efa_conn.
+ * If shm transfer is enabled and the addr comes from local peer,
+ *  1. convert addr to format 'gid_qpn', which will be set as shm's ep name later.
+ *  2. insert gid_qpn into shm's av
+ *  3. store returned fi_addr from shm into the hash table
+ *
+ * @param[in]	av	address vector
+ * @param[in]	conn	efa_conn object
+ * @return	On success return 0, otherwise return a negative error code
+ */
+static
+int efa_conn_rdm_init(struct efa_av *av, struct efa_conn *conn)
+{
+	int err, ret;
+	char smr_name[NAME_MAX];
+	struct rxr_ep *rxr_ep;
+	struct rdm_peer *peer;
+
+	assert(av->ep_type == FI_EP_RDM);
+
+	/* currently multiple EP bind to same av is not supported */
+	rxr_ep = container_of(av->util_av.ep_list.next, struct rxr_ep, util_ep.av_entry);
+
+	peer = &conn->rdm_peer;
+	ofi_atomic_initialize32(&peer->use_cnt, 1);
+	peer->efa_fiaddr = conn->fi_addr;
+	peer->is_self = efa_is_same_addr((struct efa_ep_addr *)rxr_ep->core_addr,
+					 &conn->ep_addr);
+
+	/* If peer is local, insert the address into shm provider's av */
+	if (rxr_ep->use_shm && efa_is_local_peer(av, &conn->ep_addr)) {
+		if (av->shm_used >= rxr_env.shm_av_size) {
+			EFA_WARN(FI_LOG_AV,
+				 "Max number of shm AV entry (%d) has been reached.\n",
+				 rxr_env.shm_av_size);
+			return -FI_ENOMEM;
+		}
+
+		err = rxr_ep_efa_addr_to_str(&conn->ep_addr, smr_name);
+		if (err != FI_SUCCESS) {
+			EFA_WARN(FI_LOG_AV,
+				 "rxr_ep_efa_addr_to_str() failed! err=%d\n", err);
+			return err;
+		}
+
+		ret = fi_av_insert(av->shm_rdm_av, smr_name, 1, &peer->shm_fiaddr, 0, NULL);
+		if (OFI_UNLIKELY(ret != 1)) {
+			EFA_WARN(FI_LOG_AV,
+				 "Failed to insert address to shm provider's av: %s\n",
+				 fi_strerror(-ret));
+			return err;
+		}
+
+		EFA_INFO(FI_LOG_AV,
+			"Successfully inserted %s to shm provider's av. efa_fiaddr: %ld shm_fiaddr = %ld\n",
+			smr_name, conn->fi_addr, peer->shm_fiaddr);
+
+		assert(peer->shm_fiaddr < rxr_env.shm_av_size);
+		av->shm_used++;
+		av->shm_rdm_addr_map[peer->shm_fiaddr] = conn->fi_addr;
+		peer->is_local = 1;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief release the rdm related resources of an efa_conn object
+ *
+ * this function release the shm av entry and rdm peer;
+ *
+ * @param[in]	av	address vector
+ * @param[in]	conn	efa_conn object
+ */
+static
+void efa_conn_rdm_deinit(struct efa_av *av, struct efa_conn *conn)
+{
```

Comment:
I ask because this is not just uninitializing/deinitializing ... it's removing an address from the shared memory AV.

Reply:
Because this function is correspond to `efa_conn_rdm_init`. 

`xxx_release` is used to correspond to ```xxx_alloc```, which will release the input pointer.